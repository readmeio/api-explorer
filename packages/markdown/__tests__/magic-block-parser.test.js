const unified = require('unified');
const remarkParse = require('remark-parse');
const rehypeSanitize = require('rehype-sanitize');

const parser = require('../processor/parse/magic-block-parser');
const options = require('../processor/options.json').markdownOptions;

const sanitize = { attributes: [] };
const process = (text, opts = options) =>
  text &&
  unified()
    .use(remarkParse, opts)
    .data('settings', { position: false })
    .use(parser.sanitize(sanitize))
    .use(rehypeSanitize)
    .parse(text);

test('Sanitize Schema', () => {
  parser.sanitize(sanitize);
  expect(sanitize).toMatchSnapshot();
});

describe('Parse Magic Blocks', () => {
  it('Heading Blocks', () => {
    const text = `[block:api-header]
    {
      "title": "MagicBlock Conversion",
      "level": 2
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });

  it('Image Blocks', () => {
    const text = `[block:image]
    {
      "images": [
        {
          "image": [
            "https://files.readme.io/62083ee-White_Center_Blue_BG.svg",
            "White Center Blue BG.svg",
            240,
            150,
            "#000000"
            ],
          "caption": "Qui iusto fugiat doloremque? Facilis obcaecati vitae corrupti.",
          "sizing": "80",
          "border": true
        }
      ]
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });

  it('Code Blocks', () => {
    const text = `[block:code]
    {
      "codes": [
        {
          "code": "$http.post('/someUrl', data).success(successCallback);\\n\\nalert('test');",
          "language": "javascript"
        },
        {
          "code": "second tab",
          "language": "text",
          "name": "custom title"
        }
      ]
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });

  it('Table Blocks', () => {
    const text = `[block:parameters]
    {
      "data": {
        "h-0": "Left",
        "h-1": "Center",
        "h-2": "Right",
        "0-0": "L0",
        "0-1": "*bold*",
        "0-2": "$1600",
        "2-0": "L2",
        "1-0": "L1",
        "1-1": "\`code\`",
        "1-2": "$12",
        "2-1": "_italic_",
        "2-2": "$1"
      },
      "cols": 3,
      "rows": 3
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });

  it('Embed Blocks', () => {
    const text = `[block:embed]
    {
      "html": false,
      "url": "https://youtu.be/J3-uKv1DShQ",
      "title": null,
      "favicon": "https://youtu.be/favicon.ico",
      "iframe": false
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });

  it('Callout Blocks', () => {
    const text = `[block:callout]
    {
      "type": "success",
      "title": "Callout Title",
      "body": "Lorem ipsum dolor sit amet, _consectetur_ adipiscing elit. Praesent nec massa tristique arcu fermentum dapibus. Integer orci turpis, mollis vel augue eget, placerat rhoncus orci. Mauris metus libero, rutrum"
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });

  it('Unrecognized Blocks', () => {
    const text = `[block:unrecognized]
    {
      "color": "#f00",
      "title": "Title",
      "body": "Lorem ipsum dolor sit amet, _consectetur_ adipiscing elit. Praesent nec massa tristique arcu fermentum dapibus. Integer orci turpis, mollis vel augue eget, placerat rhoncus orci. Mauris metus libero, rutrum"
    }
    [/block]`;
    expect(process(text)).toMatchSnapshot();
  });
});
