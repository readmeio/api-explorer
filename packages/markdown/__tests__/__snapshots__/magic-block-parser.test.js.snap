// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Parse Magic Blocks Callout Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "üëç ",
            },
            Object {
              "type": "text",
              "value": "Callout Title",
            },
          ],
          "type": "paragraph",
        },
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Lorem ipsum dolor sit amet, ",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "consectetur",
                },
              ],
              "type": "emphasis",
            },
            Object {
              "type": "text",
              "value": " adipiscing elit. Praesent nec massa tristique arcu fermentum dapibus. Integer orci turpis, mollis vel augue eget, placerat rhoncus orci. Mauris metus libero, rutrum",
            },
          ],
          "type": "paragraph",
        },
      ],
      "data": Object {
        "hName": "rdme-callout",
        "hProperties": Object {
          "icon": "üëç",
          "theme": "okay",
          "title": "Callout Title",
          "value": "Lorem ipsum dolor sit amet, _consectetur_ adipiscing elit. Praesent nec massa tristique arcu fermentum dapibus. Integer orci turpis, mollis vel augue eget, placerat rhoncus orci. Mauris metus libero, rutrum",
        },
      },
      "type": "rdme-callout",
    },
  ],
  "type": "root",
}
`;

exports[`Parse Magic Blocks Code Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "className": "tab-panel",
          "data": Object {
            "hName": "code",
            "hProperties": Object {
              "lang": "javascript",
              "meta": null,
            },
          },
          "lang": "javascript",
          "meta": null,
          "type": "code",
          "value": "$http.post('/someUrl', data).success(successCallback);

alert('test');",
        },
        Object {
          "className": "tab-panel",
          "data": Object {
            "hName": "code",
            "hProperties": Object {
              "lang": "text",
              "meta": "custom title",
            },
          },
          "lang": "text",
          "meta": "custom title",
          "type": "code",
          "value": "second tab",
        },
      ],
      "className": "tabs",
      "data": Object {
        "hName": "code-tabs",
      },
      "type": "code-tabs",
    },
  ],
  "type": "root",
}
`;

exports[`Parse Magic Blocks Embed Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "type": "text",
              "value": "Embed",
            },
          ],
          "title": "@youtube",
          "type": "link",
          "url": "https://youtu.be/J3-uKv1DShQ",
        },
      ],
      "data": Object {
        "hName": "rdme-embed",
        "hProperties": Object {
          "href": "https://youtu.be/J3-uKv1DShQ",
          "html": false,
          "provider": "@youtube",
          "title": "Embed",
          "url": "https://youtu.be/J3-uKv1DShQ",
        },
        "html": false,
        "provider": "@youtube",
        "title": "Embed",
        "url": "https://youtu.be/J3-uKv1DShQ",
      },
      "html": false,
      "provider": "@youtube",
      "title": "Embed",
      "type": "embed",
      "url": "https://youtu.be/J3-uKv1DShQ",
    },
  ],
  "type": "root",
}
`;

exports[`Parse Magic Blocks Heading Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "type": "text",
          "value": "MagicBlock Conversion",
        },
      ],
      "depth": 2,
      "type": "heading",
    },
  ],
  "type": "root",
}
`;

exports[`Parse Magic Blocks Image Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "alt": "Qui iusto fugiat doloremque? Facilis obcaecati vitae corrupti.",
      "data": Object {
        "hProperties": Object {
          "caption": "Qui iusto fugiat doloremque? Facilis obcaecati vitae corrupti.",
        },
      },
      "title": "White Center Blue BG.svg",
      "type": "image",
      "url": "https://files.readme.io/62083ee-White_Center_Blue_BG.svg",
    },
  ],
  "type": "root",
}
`;

exports[`Parse Magic Blocks Table Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "align": Array [
        "left",
        "left",
        "left",
      ],
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Left",
                },
              ],
              "type": "tableHead",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Center",
                },
              ],
              "type": "tableHead",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "Right",
                },
              ],
              "type": "tableHead",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "L0",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "bold",
                    },
                  ],
                  "type": "emphasis",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "$1600",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "L1",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "inlineCode",
                  "value": "code",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "$12",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "L2",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "type": "text",
                      "value": "italic",
                    },
                  ],
                  "type": "emphasis",
                },
              ],
              "type": "tableCell",
            },
            Object {
              "children": Array [
                Object {
                  "type": "text",
                  "value": "$1",
                },
              ],
              "type": "tableCell",
            },
          ],
          "type": "tableRow",
        },
      ],
      "type": "table",
    },
  ],
  "type": "root",
}
`;

exports[`Parse Magic Blocks Unrecognized Blocks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [],
      "data": Object {
        "body": "Lorem ipsum dolor sit amet, _consectetur_ adipiscing elit. Praesent nec massa tristique arcu fermentum dapibus. Integer orci turpis, mollis vel augue eget, placerat rhoncus orci. Mauris metus libero, rutrum",
        "color": "#f00",
        "title": "Title",
      },
      "type": "div",
    },
  ],
  "type": "root",
}
`;

exports[`Sanitize Schema 1`] = `
Object {
  "attributes": Array [],
}
`;
